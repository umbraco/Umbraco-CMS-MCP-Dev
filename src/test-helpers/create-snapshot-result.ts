import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
