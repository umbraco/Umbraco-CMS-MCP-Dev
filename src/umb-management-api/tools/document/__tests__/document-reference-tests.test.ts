import GetDocumentAreReferencedTool from "../get/get-document-are-referenced.js";
import GetDocumentByIdReferencedByTool from "../get/get-document-by-id-referenced-by.js";
import GetDocumentByIdReferencedDescendantsTool from "../get/get-document-by-id-referenced-descendants.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { DocumentBuilder } from "./helpers/document-builder.js";
import { DocumentTestHelper } from "./helpers/document-test-helper.js";
import { jest } from "@jest/globals";

const TEST_DOCUMENT_NAME = "_Test Reference Document";
const TEST_DOCUMENT_NAME_2 = "_Test Reference Document 2";

describe("document-reference-tests", () => {
  let originalConsoleError: typeof console.error;

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
  });

  afterEach(async () => {
    console.error = originalConsoleError;
    await DocumentTestHelper.cleanup(TEST_DOCUMENT_NAME);
    await DocumentTestHelper.cleanup(TEST_DOCUMENT_NAME_2);
  });

  describe("get-document-are-referenced", () => {
    it("should check if documents are referenced", async () => {
      // Arrange: Create documents
      const builder1 = await new DocumentBuilder()
        .withName(TEST_DOCUMENT_NAME)
        .withRootDocumentType()
        .create();

      const builder2 = await new DocumentBuilder()
        .withName(TEST_DOCUMENT_NAME_2)
        .withRootDocumentType()
        .create();

      // Act: Check if documents are referenced
      const result = await GetDocumentAreReferencedTool().handler(
        { id: [builder1.getId(), builder2.getId()], take: 20 },
        { signal: new AbortController().signal }
      );

      // Assert: Verify the response
      const normalizedResult = createSnapshotResult(result);
      expect(normalizedResult).toMatchSnapshot();
    });

    it("should handle single document reference check", async () => {
      // Arrange: Create a document
      const builder = await new DocumentBuilder()
        .withName(TEST_DOCUMENT_NAME)
        .withRootDocumentType()
        .create();

      // Act: Check if single document is referenced
      const result = await GetDocumentAreReferencedTool().handler(
        { id: [builder.getId()], skip: 0, take: 10 },
        { signal: new AbortController().signal }
      );

      // Assert: Verify the response
      const normalizedResult = createSnapshotResult(result);
      expect(normalizedResult).toMatchSnapshot();
    });

    it("should handle empty reference check", async () => {
      // Act: Check references for empty array
      const result = await GetDocumentAreReferencedTool().handler(
        { id: [], take: 20 },
        { signal: new AbortController().signal }
      );

      // Assert: Should handle gracefully
      expect(result).toMatchSnapshot();
    });
  });

  describe("get-document-by-id-referenced-by", () => {
    it("should get documents that reference a document by ID", async () => {
      // Arrange: Create a document
      const builder = await new DocumentBuilder()
        .withName(TEST_DOCUMENT_NAME)
        .withRootDocumentType()
        .create();

      // Act: Get documents that reference this document
      const result = await GetDocumentByIdReferencedByTool().handler(
        { id: builder.getId(), take: 20 },
        { signal: new AbortController().signal }
      );

      // Assert: Verify the response
      const normalizedResult = createSnapshotResult(result);
      expect(normalizedResult).toMatchSnapshot();
    });

    it("should get references with pagination", async () => {
      // Arrange: Create a document
      const builder = await new DocumentBuilder()
        .withName(TEST_DOCUMENT_NAME)
        .withRootDocumentType()
        .create();

      // Act: Get references with pagination
      const result = await GetDocumentByIdReferencedByTool().handler(
        { id: builder.getId(), skip: 0, take: 10 },
        { signal: new AbortController().signal }
      );

      // Assert: Verify the response
      const normalizedResult = createSnapshotResult(result);
      expect(normalizedResult).toMatchSnapshot();
    });

    it("should handle non-existent document", async () => {
      // Act: Try to get references for non-existent document
      const result = await GetDocumentByIdReferencedByTool().handler(
        { id: "00000000-0000-0000-0000-000000000000", take: 20 },
        { signal: new AbortController().signal }
      );

      // Assert: Should handle gracefully
      expect(result).toMatchSnapshot();
    });
  });

  describe("get-document-by-id-referenced-descendants", () => {
    it("should get descendants that are referenced", async () => {
      // Arrange: Create a document
      const builder = await new DocumentBuilder()
        .withName(TEST_DOCUMENT_NAME)
        .withRootDocumentType()
        .create();

      // Act: Get referenced descendants
      const result = await GetDocumentByIdReferencedDescendantsTool().handler(
        { id: builder.getId(), take: 20 },
        { signal: new AbortController().signal }
      );

      // Assert: Verify the response
      const normalizedResult = createSnapshotResult(result);
      expect(normalizedResult).toMatchSnapshot();
    });

    it("should get referenced descendants with pagination", async () => {
      // Arrange: Create a document
      const builder = await new DocumentBuilder()
        .withName(TEST_DOCUMENT_NAME)
        .withRootDocumentType()
        .create();

      // Act: Get referenced descendants with pagination
      const result = await GetDocumentByIdReferencedDescendantsTool().handler(
        { id: builder.getId(), skip: 0, take: 10 },
        { signal: new AbortController().signal }
      );

      // Assert: Verify the response
      const normalizedResult = createSnapshotResult(result);
      expect(normalizedResult).toMatchSnapshot();
    });

    it("should handle non-existent document", async () => {
      // Act: Try to get referenced descendants for non-existent document
      const result = await GetDocumentByIdReferencedDescendantsTool().handler(
        { id: "00000000-0000-0000-0000-000000000000", take: 20 },
        { signal: new AbortController().signal }
      );

      // Assert: Should handle gracefully
      expect(result).toMatchSnapshot();
    });
  });
});