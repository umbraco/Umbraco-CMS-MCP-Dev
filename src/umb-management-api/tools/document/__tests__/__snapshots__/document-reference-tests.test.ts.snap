// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document-reference-tests get-document-are-referenced should check if documents are referenced 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-are-referenced should handle empty reference check 1`] = `
{
  "content": [
    {
      "text": "Error using get-document-are-referenced:
{
  "message": "Request failed with status code 500",
  "response": {
    "type": "Error",
    "title": "Incorrect syntax near ')'.",
    "status": 500,
    "detail": "   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, SqlCommand command, Boolean callerHasConnectionLock, Boolean asyncClose)\\n   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)\\n   at Microsoft.Data.SqlClient.SqlDataReader.TryConsumeMetaData()\\n   at Microsoft.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)\\n   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean isAsync, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)\\n   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, TaskCompletionSource\`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String method)\\n   at Microsoft.Data.SqlClient.SqlCommand.ExecuteScalar()\\n   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.RetryPolicy.ExecuteAction[TResult](Func\`1 func)\\n   at NPoco.Database.ExecuteScalarHelper(DbCommand cmd)\\n   at NPoco.Database.ExecuteScalar[T](String sql, CommandType commandType, Object[] args)\\n   at Umbraco.Cms.Infrastructure.Persistence.UmbracoDatabase.ExecuteScalar[T](String sql, CommandType commandType, Object[] args)\\n   at Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement.TrackedReferencesRepository.GetPagedNodeKeysWithDependantReferencesAsync(ISet\`1 keys, Guid nodeObjectTypeId, Int64 skip, Int64 take)\\n   at Umbraco.Cms.Core.Services.TrackedReferencesService.GetPagedKeysWithDependentReferencesAsync(ISet\`1 keys, Guid objectTypeId, Int64 skip, Int64 take)\\n   at Umbraco.Cms.Api.Management.Controllers.Document.References.AreReferencedDocumentController.GetPagedReferencedItems(CancellationToken cancellationToken, HashSet\`1 ids, Int32 skip, Int32 take)\\n   at lambda_method89332(Closure, Object)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask\`1 actionResultValueTask)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()\\n--- End of stack trace from previous location ---\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()\\n--- End of stack trace from previous location ---\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)\\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)\\n   at Umbraco.Cms.Web.Common.Middleware.BasicAuthenticationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.InterfaceMiddlewareBinder.<>c__DisplayClass2_0.<<CreateMiddleware>b__0>d.MoveNext()\\n--- End of stack trace from previous location ---\\n   at Umbraco.Cms.Api.Management.Middleware.BackOfficeExternalLoginProviderErrorMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.InterfaceMiddlewareBinder.<>c__DisplayClass2_0.<<CreateMiddleware>b__0>d.MoveNext()\\n--- End of stack trace from previous location ---\\n   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)\\n   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)\\n   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)\\n   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)\\n   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)\\n   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)\\n   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)\\n   at StackExchange.Profiling.MiniProfilerMiddleware.Invoke(HttpContext context) in C:\\\\projects\\\\dotnet\\\\src\\\\MiniProfiler.AspNetCore\\\\MiniProfilerMiddleware.cs:line 112\\n   at Umbraco.Cms.Web.Common.Middleware.UmbracoRequestMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)\\n   at Umbraco.Cms.Web.Common.Middleware.UmbracoRequestMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.InterfaceMiddlewareBinder.<>c__DisplayClass2_0.<<CreateMiddleware>b__0>d.MoveNext()\\n--- End of stack trace from previous location ---\\n   at Umbraco.Cms.Web.Common.Middleware.PreviewAuthenticationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.InterfaceMiddlewareBinder.<>c__DisplayClass2_0.<<CreateMiddleware>b__0>d.MoveNext()\\n--- End of stack trace from previous location ---\\n   at Umbraco.Cms.Web.Common.Middleware.UmbracoRequestLoggingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)\\n   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.InterfaceMiddlewareBinder.<>c__DisplayClass2_0.<<CreateMiddleware>b__0>d.MoveNext()\\n--- End of stack trace from previous location ---\\n   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddlewareImpl.<Invoke>g__Awaited|10_0(ExceptionHandlerMiddlewareImpl middleware, HttpContext context, Task task)",
    "instance": "SqlException"
  }
}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-are-referenced should handle single document reference check 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-by-id-referenced-by should get documents that reference a document by ID 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-by-id-referenced-by should get references with pagination 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-by-id-referenced-by should handle non-existent document 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-by-id-referenced-descendants should get descendants that are referenced 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-by-id-referenced-descendants should get referenced descendants with pagination 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;

exports[`document-reference-tests get-document-by-id-referenced-descendants should handle non-existent document 1`] = `
{
  "content": [
    {
      "text": "{"total":0,"items":[]}",
      "type": "text",
    },
  ],
}
`;
